File: exporters/gcp_to_cwl/pubsub.js
const google = require('googleapis');

// TODO smoke test
exports.SynchronousPullClient = function(options) {
    this.hostname = 'pubsub.googleapis.com';
    this.projectId = options.projectId;
    this.subscription = options.subscription;
}


exports.SynchronousPullClient.prototype.auth = function() {
  return new Promise(function(resolve, reject) {
    google.auth.getApplicationDefault(function(err, authClient, projectId) {
        if (err) {
            reject(err);
            return;
        }
        if (authClient.createScopedRequired
            && authClient.createScopedRequired()
           ) {
            const scopes = [
                'https://www.googleapis.com/auth/cloud-platform',
                'https://www.googleapis.com/auth/pubsub',
            ];
            authClient = authClient.createScoped(scopes);
        }

        resolve(authClient);
    });
  });
}

exports.SynchronousPullClient.prototype.queryPath = function(operation) {
    return `v1/projects/${this.projectId}/subscriptions/${this.subscription}:${operation}`
}

exports.SynchronousPullClient.prototype.pull = function(size, callback) {
  var postData = {returnImmediately: false, maxMessages: size};
  const options = {
    url: `https://${this.hostname}/${this.queryPath('pull')}`,
    method: 'POST',
    json: true,
    body: postData,
  }
  return this.auth().then(authClient => {
    return authClient.request(options, function(err, body, response) {
      if (!err) return callback(body)
      throw err
    });
  });
}

exports.SynchronousPullClient.prototype.ack = function(ackIds, callback) {
  if (ackIds.length == 0) return callback({});
  var postData = {ackIds: ackIds};
  const options = {
    url: `https://${this.hostname}/${this.queryPath('acknowledge')}`,
    method: 'POST',
    json: true,
    body: postData,
  }
  return this.auth().then(authClient => {
    return authClient.request(options, function(err, body, response) {
      if (!err) return callback(body)
      else throw err
    });
  });
}


exports.SynchronousPullClient.prototype.stream = function(n, processCallback, endCallback) {
  return this.pull(n, body => {
    var [messages, ackIds] = exports.openSynchronousPullEnvelope(body);
    return processCallback(messages)
      .then(callbackResult => this.ack(ackIds, responseBody => callbackResult))
      .then(ackResult => {
        if (messages.length == n) {
          return this.stream(n, processCallback, endCallback);
        }
        return endCallback();
      })
  })
}


var decode = function(data) {
  return JSON.parse(Buffer.from(data, 'base64').toString());
}


exports.openSynchronousPullEnvelope = function(envelope) {
  if (envelope.receivedMessages == undefined) return [[], []];
  var ackIds = envelope.receivedMessages.map(m => m.ackId);
  var messages = envelope.receivedMessages.map(m => decode(m.message.data));
  return [messages, ackIds];
}

File: exporters/gcp_to_cwl/domovoilib/pubsub.py
import json
import typing
from google.cloud import pubsub_v1
from google.cloud.pubsub_v1.proto import pubsub_pb2


class SynchronousPullClient:

    def __init__(self, project_id, subscription_name):
        self.client = pubsub_v1.SubscriberClient()
        self.subscription = self.client.subscription_path(project_id, subscription_name)

    @staticmethod
    def _open_pull_envelope(envelope) -> typing.Tuple[typing.List[dict], typing.List[str]]:
        ack_ids = [m.ack_id for m in envelope.received_messages]
        messages = [
            json.loads(str(m.message.data, 'utf-8'))
            for m in envelope.received_messages
        ]
        assert(len(ack_ids) == len(messages))
        return messages, ack_ids

    def pull(self, batch_size) -> typing.Tuple[typing.List[dict], typing.List[str]]:
        request = pubsub_pb2.PullRequest(
            subscription=self.subscription,
            max_messages=batch_size,
            return_immediately=True)
        response = self.client.api._pull(request)
        return self._open_pull_envelope(response)

    def ack(self, ack_ids):
        self.client.acknowledge(self.subscription, ack_ids)

    def to_generator(self, batch_size):
        low_poll_count = 0
        while low_poll_count < 2:
            messages, ack_ids = self.pull(batch_size)
            if len(messages) < batch_size:
                low_poll_count += 1
            else:
                low_poll_count = 0

            if len(messages) > 0:
                yield messages
                self.ack(ack_ids)

